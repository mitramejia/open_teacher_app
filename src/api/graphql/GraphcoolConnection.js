// =============================================================
//
// Name: GraphcoolConnection
// -> Description:
//
// Author: mitramejia
// Created at: 8/21/17
//
// =============================================================

import { ApolloClient, createNetworkInterface } from 'apollo-client';
import { GRAPHCOOL_SIMPLE_API, GRAPHCOOL_WEB_SOCKET_CLIENT } from 'react-native-dotenv';
import { addGraphQLSubscriptions, SubscriptionClient } from 'subscriptions-transport-ws';
import { AsyncStorage } from 'react-native';

const networkInterface = createNetworkInterface({ uri: GRAPHCOOL_SIMPLE_API });

// Use middleware to manage auth tokens
networkInterface.use([
  {
    applyMiddleware(req, next) {
      if (!req.options.headers) {
        req.options.headers = {}; // Create the header object if needed.
      }

      AsyncStorage.getItem('token').then(
        encodedToken => {
          req.options.headers['authorization'] = `Bearer ${encodedToken}`;
          next();
        },
        failure => {
          console.error('ERROR: no token', failure);
          next();
        }
      );
    },
  },
]);

// Create WebSocket client
const wsClient = new SubscriptionClient(GRAPHCOOL_WEB_SOCKET_CLIENT, {
  reconnect: true,
});
// Extend the network interface with the WebSocket
const networkInterfaceWithSubscriptions = addGraphQLSubscriptions(networkInterface, wsClient);

const client = new ApolloClient({
  networkInterface: networkInterfaceWithSubscriptions,
  dataIdFromObject: o => o.id,
});

export default client;
