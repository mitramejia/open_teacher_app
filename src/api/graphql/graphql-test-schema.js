// =============================================================
//
// Name: graphql-test-schema
// -> Description:
//
// Author: mitramejia
// Created at: 8/28/17
//
// =============================================================

import { makeExecutableSchema, addMockFunctionsToSchema } from 'graphql-tools';
import { mockNetworkInterfaceWithSchema } from 'apollo-test-utils';
import { ApolloClient } from 'apollo-client';

export function mockApolloClient() {
  const schema = makeExecutableSchema({ typeDefs });
  addMockFunctionsToSchema({ schema });
  const mockNetworkInterface = mockNetworkInterfaceWithSchema({ schema });
  return new ApolloClient({
    networkInterface: mockNetworkInterface,
  });
}

export const typeDefs = `
      type User {
        id: Int
        firstName: String
        lastName: String
        email: String
      }

      type Query {
        user: User
      }
    `;
