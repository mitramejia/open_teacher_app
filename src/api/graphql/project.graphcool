# project: cj5zswpmnfr6201049umywa1m
# version: 48

type Post implements Node {
  description: String!
  imageUrl: String!
  createdAt: DateTime!
  id: ID! @isUnique
  updatedAt: DateTime!
}

type File implements Node {
  contentType: String!
  createdAt: DateTime!
  id: ID! @isUnique
  name: String!
  secret: String! @isUnique
  size: Int!
  updatedAt: DateTime!
  url: String! @isUnique
}

type Tutor implements Node {
  createdAt: DateTime!
  id: ID! @isUnique
  updatedAt: DateTime!
  students: [Student!]! @relation(name: "TutorOnStudent")
  description: String
  diplomaUrl: String!
  lessons: [Lesson!]! @relation(name: "TutorOnLesson")
  locations: [String!]!
  subjects: [Subject!]! @relation(name: "SubjectOnTutor")
  user: User! @relation(name: "UserOnTutor")
}

type User implements Node {
  createdAt: DateTime!
  firstName: String!
  imageUrl: String
  lastName: String!
  id: ID! @isUnique
  updatedAt: DateTime!
  email: String @isUnique
  isTutor: Boolean! @defaultValue(value: false)
  password: String
  student: Student @relation(name: "UserOnStudent")
  tutor: Tutor @relation(name: "UserOnTutor")
}

type Lesson implements Node {
  cost: Float!
  createdAt: DateTime!
  duration: Int!
  id: ID! @isUnique
  student: Student @relation(name: "StudentOnLesson")
  updatedAt: DateTime!
  user: Tutor @relation(name: "TutorOnLesson")
}

type Student implements Node {
  createdAt: DateTime!
  id: ID! @isUnique
  lessons: [Lesson!]! @relation(name: "StudentOnLesson")
  subjects: [Subject!]! @relation(name: "SubjectOnStudent")
  tutor: Tutor @relation(name: "TutorOnStudent")
  updatedAt: DateTime!
  user: User! @relation(name: "UserOnStudent")
}

type Subject implements Node {
  createdAt: DateTime!
  id: ID! @isUnique
  name: String!
  students: [Student!]! @relation(name: "SubjectOnStudent")
  tutors: [Tutor!]! @relation(name: "SubjectOnTutor")
  updatedAt: DateTime!
}