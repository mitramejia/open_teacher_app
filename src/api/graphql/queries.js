// =============================================================
//
// Name: model
// -> Description:
//
// Author: mitramejia
// Created at: 8/25/17
//
// =============================================================

import { gql } from 'react-apollo';

export const createAccount = gql`
  mutation($firstName: String!, $lastName: String!, $email: String!, $password: String!) {
    createUser(
      authProvider: { email: { email: $email, password: $password } }
      firstName: $firstName
      lastName: $lastName
    ) {
      id
      firstName
      lastName
      email
      password
    }
  }
`;

export const userQuery = gql`
  query {
    user {
      id
      firstName
      lastName
      email
    }
  }
`;

export const signinUser = gql`
  mutation($email: String!, $password: String!) {
    signinUser(email: { email: $email, password: $password }) {
      token
    }
  }
`;

export const allSubjectsQuery = gql`
  query {
    allSubjects {
      id
      name
    }
  }
`;
