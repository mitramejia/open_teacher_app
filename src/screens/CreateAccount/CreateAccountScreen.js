// =============================================================
//
// Name: StudentHomeScreen
// -> Description:
//
// Author: mitramejia
// Created at: 8/21/17
//
// =============================================================

import React from 'react';
import { AppLoading } from 'expo';
import { Content, Button, Text, Item, Toast } from 'native-base';
import { Keyboard, TextInput, AsyncStorage } from 'react-native';
import { compose, gql, graphql, withApollo } from 'react-apollo';
import _ from 'lodash';
import propTypes from 'prop-types';
import {
  Form,
  Separator,
  InputField,
  LinkField,
  SwitchField,
  PickerField,
  DatePickerField,
  TimePickerField,
} from 'react-native-form-generator';
import { Ionicons } from '@expo/vector-icons';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import style from './style';
import { createAccount, signinUser, userQuery } from '../../api/graphql/queries';

class CreateAccountScreen extends React.Component {
  constructor(...props) {
    super(...props);
    this.state = {
      showToast: false,
      formData: {
        firstName: '',
        lastName: '',
        email: '',
        password: '',
      },
      formError: true,
    };
  }

  static navigationOptions = {
    title: 'Crear Cuenta',
  };

  _handleFormChange(formData) {
    this.setState({ formData });
    this.props.onFormChange && this.props.onFormChange(formData);
  }

  _createAccount() {
    const { firstName, lastName, email, password } = this.state.formData;
    this.props
      .createAccount({ variables: { firstName, lastName, email, password } })
      .then(response => {
        console.log(response);
        this.props
          .signinUser({ variables: { email, password } })
          .then(response => {
            console.log(response);
            AsyncStorage.setItem('token', response.data.signinUser.token);
            this.props.navigation.navigate('Home');
            Toast.show({
              text: `Hola ${firstName}`,
              position: 'bottom',
              buttonText: 'Ocultar',
              type: 'success',
              duration: 4000,
            });
          })
          .catch(e => {
            console.error(e.message);
            this.props.navigation.navigate('Home');
          });
      })
      .catch(e => {
        console.error(e.message);
        this.props.navigation.navigate('Home');
      });
    console.log('User: ' + this.props.data.user);
  }

  _checkAllFieldsAreValid() {
    return _.every(this.state.formData, value => {
      // Check all form fields aren't empty or undefined
      if (value !== '' || value !== undefined) {
        return true;
      }
    });
  }

  render() {
    if (this.props.data.loading) {
      return <AppLoading />;
    } else {
      return (
        <KeyboardAwareScrollView
          contentContainerStyle={style.container}
          keyboardShouldPersistTaps="always"
          ref="scroll">
          <Form
            ref="registrationForm"
            onChange={this._handleFormChange.bind(this)}
            label="Create Account">
            <InputField style={style.inputField} ref="firstName" label="Nombre" />
            <InputField style={style.inputField} ref="lastName" label="Apellido" />
            <InputField
              style={style.inputField}
              ref="email"
              label="Email"
              keyboardType="email-address"
            />
            <InputField
              style={style.inputField}
              ref="password"
              label="ContraseÃ±a"
              secureTextEntry={true}
            />
          </Form>
          {this._checkAllFieldsAreValid() &&
            <Button
              rounded
              block
              style={style.submitButton}
              onPress={this._createAccount.bind(this)}>
              <Text>Crear Cuenta</Text>
            </Button>}
          <Text>
            {JSON.stringify(this.state.formData)}
          </Text>
        </KeyboardAwareScrollView>
      );
    }
  }
}

CreateAccountScreen.propTypes = {
  client: React.PropTypes.object.isRequired,
  navigation: React.PropTypes.object.isRequired,
  createAccount: React.PropTypes.func.isRequired,
  signinUser: React.PropTypes.func.isRequired,
  data: React.PropTypes.object.isRequired,
};

export default graphql(createAccount, { name: 'createAccount' })(
  graphql(userQuery, { options: { fetchPolicy: 'network-only' } })(
    graphql(signinUser, { name: 'signinUser' })(withApollo(CreateAccountScreen))
  )
);
