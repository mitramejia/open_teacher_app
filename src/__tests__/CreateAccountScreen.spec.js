import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { ApolloProvider } from 'react-apollo';
import { mockApolloClient } from '../api/graphql/graphql-test-schema';
import CreateAccountScreen from '../screens/CreateAccount/CreateAccountScreen';

it('renders a Create Account Screen', () => {
  const navigation = {};
  const user = {};
  const tree = renderer
    .create(
      <ApolloProvider client={mockApolloClient()}>
        <CreateAccountScreen navigation={navigation} user={user} />
      </ApolloProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
