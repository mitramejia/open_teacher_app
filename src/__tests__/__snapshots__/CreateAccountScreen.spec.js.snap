// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Create Account Screen 1`] = `<ExponentAppLoading />`;
